[GEN][ZH] Fix incorrect string copy with overlapping memory regions in AsciiString, UnicodeString, strtrim

Mauller opened the pull request:
The two commits here correct some functionality that was causing ASAN to trigger on startup.

The source and destination buffers, according to ASAN, have overlapping memory regions meaning the copy would be unsafe to perform.

The solution and safe way to do this, if it was intended, such as shuffling characters along in a buffer, you should use memmove.
memmove is used in other parts of the code base as well so has compat with VC6.

To see this you must have the null memory implementation in use and ASAN running.

EDIT: Updated to include Unicode and wide character string trim.

Mauller commented:
Example callstack being thrown, this occurs in the same place in both games.

```CPP
 	KernelBase.dll!750220c2()	Unknown
 	[Frames below may be incorrect and/or missing, no symbols loaded for KernelBase.dll]	
 	[External Code]	
>	generalszh.exe!AsciiString::ensureUniqueBufferOfSize(int numCharsNeeded, bool preserveData, const char * strToCopy, const char * strToCat) Line 130	C++
 	generalszh.exe!AsciiString::set(const char * s) Line 480	C++
 	generalszh.exe!AsciiString::nextToken(AsciiString * tok, const char * seps) Line 388	C++
 	generalszh.exe!ArchiveFile::addFile(const AsciiString & path, const ArchivedFileInfo * fileInfo) Line 107	C++
 	generalszh.exe!Win32BIGFileSystem::openArchiveFile(const char * filename) Line 175	C++
 	generalszh.exe!Win32BIGFileSystem::loadBigFilesFromDirectory(AsciiString dir, AsciiString fileMask, bool overwrite) Line 222	C++
 	generalszh.exe!Win32BIGFileSystem::init() Line 61	C++
 	generalszh.exe!SubsystemInterfaceList::initSubsystem(SubsystemInterface * sys, const char * path1, const char * path2, const char * dirpath, Xfer * pXfer, AsciiString name) Line 164	C++
 	generalszh.exe!initSubsystem<ArchiveFileSystem>(ArchiveFileSystem * & sysref, AsciiString name, ArchiveFileSystem * sys, Xfer * pXfer, const char * path1, const char * path2, const char * dirpath) Line 167	C++
 	generalszh.exe!GameEngine::init(int argc, char * * argv) Line 353	C++
 	generalszh.exe!GameMain(int argc, char * * argv) Line 44	C++
 	generalszh.exe!WinMain(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) Line 1067	C++
 	[External Code]	
```

inside the function it is then calling.
```C++
		if (strToCopy)
			strcpy(m_data->peek(), strToCopy);
```

xezon commented on Generals/Code/GameEngine/Source/Common/System/AsciiString.cpp:
```diff
@@ -131,7 +131,8 @@ void AsciiString::ensureUniqueBufferOfSize(int numCharsNeeded, Bool preserveData
 		if (strToCopy)
-			strcpy(m_data->peek(), strToCopy);
+			// TheSuperHackers @fix Mauller 04/04/2025 ASAN, Replace strcpy with safer memmove as copy occurs within overlapping address space
+			memmove(m_data->peek(), strToCopy, strlen(strToCopy) + 1);
```
Does this fix an observable bug? If not, then this is a minor fix.

Mauller replied:
I think under normal use it isn't observable since strcpy will still perform what is required, but memmove can perform it in a safer way when memory regions do overlap.

Mauller replied:
updated

xezon commented on Generals/Code/GameEngine/Source/Common/System/AsciiString.cpp:
```diff
@@ -131,7 +131,8 @@ void AsciiString::ensureUniqueBufferOfSize(int numCharsNeeded, Bool preserveData
 		// no buffer manhandling is needed (it's already large enough, and unique to us)
 		if (strToCopy)
-			strcpy(m_data->peek(), strToCopy);
+			// TheSuperHackers @fix Mauller 04/04/2025 ASAN, Replace strcpy with safer memmove as copy occurs within overlapping address space
```
No need to mention asan here.

"Replaces strcpy with memmove because copy can occur within overlapping address space"

xezon commented on Generals/Code/Libraries/Source/WWVegas/WWLib/trim.cpp:
```diff
@@ -71,7 +71,8 @@ char * strtrim(char * buffer)
 		if (source != buffer) {
-			strcpy(buffer, source);
+			// TheSuperHackers @fix Mauller 04/04/2025 ASAN, Replace strcpy with safer memmove as copy occurs within overlapping address space
+			memmove(buffer, source, strlen(source) +1);
```
What about UnicodeString class? Does not have this bug?

xezon replied:
`UnicodeString` class also has
```cpp
		// no buffer manhandling is needed (it's already large enough, and unique to us)
		if (strToCopy)
			wcscpy(m_data->peek(), strToCopy);
```

Mauller replied:
updated

xezon commented on Generals/Code/GameEngine/Source/Common/System/AsciiString.cpp:
```diff
@@ -131,7 +131,8 @@ void AsciiString::ensureUniqueBufferOfSize(int numCharsNeeded, Bool preserveData
 	{
 		// no buffer manhandling is needed (it's already large enough, and unique to us)
 		if (strToCopy)
-			strcpy(m_data->peek(), strToCopy);
```
What about the other strcpy in this function? Shouldn't that also do memmove?
```cpp
	// do these BEFORE releasing the old buffer, so that self-copies
	// or self-cats will work correctly.
	if (strToCopy)
		strcpy(newData->peek(), strToCopy);
```

Mauller replied:
I had not fully checked back over these just yet, but that comment does smell a bit and point to it requiring a memmove instead.

The initial changes were what Asan triggered on and required to progress to the main menu.

Mauller replied:
udpated

Mauller commented:
Updated to include fixes for unicode string and the wide char version of string trim.

Mauller commented on GeneralsMD/Code/Libraries/Source/WWVegas/WWLib/trim.cpp:
```diff
@@ -68,7 +68,8 @@ char* strtrim(char* buffer)
 		if (source != buffer) {
-			strcpy(buffer, source);
+			// TheSuperHackers @fix Mauller 04/04/2025 Replace strcpy with safer memmove as memory regions can overlap when part of string is copied to itself
+			memmove(buffer, source, strlen(source) +1);
```
Ah i did not realise there were other copies laying about, maybe we should remove them and just use a single instance of it here?

Mauller replied:
i will look at removing the duplicates in this case and including the WWLIB variant of trim in relevant places.

xezon replied:
Does that mean this change is ready to go or do you first need to look around?

Mauller replied:
> Does that mean this change is ready to go or do you first need to look around?

Just fixing up the over instances of strtrim, for the winmain variants i have included trim.h and removed the local copy.

For the others i have replaced the local copy with the copy from trim, they don't include wwlib as part of their build so prob better to just fix those ones up and keep them local for now, they seem like they are in tools that will get dropped in the end.

Mauller commented:
Updated and replaced local copys with trim.h where applicable and updated the local copys for the rest with the variant from trim.cpp

Some of the builds do not include wwlib, so i just left the local copies but updated them.

xezon commented on GeneralsMD/Code/Tools/versionUpdate/versionUpdate.cpp:
```diff
@@ -92,36 +92,34 @@ static void usage(char *progname)
-			strcpy(buffer, source);
+		if (source != buffer) {
+			// TheSuperHackers @fix Mauller 04/04/2025 Replace strcpy with safer memmove as memory regions can overlap when part of string is copied to itself
+			memmove(buffer, source, strlen(source) +1);
```
This change now does more than advertised. The unification of strtrim would be better as a separate change.

Mauller replied:
Reverted it back to just the unicode, ascii and trim changes.

Mauller commented:
Reverted to just the String library and Trim changes.

Trim unification pulled into it's own PR.
